### PR TITLE
`azurerm_kubernetes_cluster` - Support in-place update of `network_profile.network_policy`

### PR DESCRIPTION
**Pull request by jkroepke**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This is an follow-up PR from https://github.com/hashicorp/terraform-provider-azurerm/pull/23342

Reading the AKS docs https://learn.microsoft.com/en-us/azure/aks/use-network-policies it's now possible to upgrade from none to calico, azure or cillium. It also possible to uninstall calico and azure.

There are now too many situation that they are allowed are now. Removing the ForceNew flag should be the best option here. 

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [ ] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containers '-run=TestAccKubernetesCluster_.*Policy.*' -timeout 300m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesCluster_addonProfileAzurePolicy
=== PAUSE TestAccKubernetesCluster_addonProfileAzurePolicy
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicy
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicy
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicy
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicy
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyComplete
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyComplete
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyNetworkModeTransparent
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyNetworkModeTransparent
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicy
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicy
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicyComplete
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicyComplete
=== CONT  TestAccKubernetesCluster_addonProfileAzurePolicy
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicyComplete
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicy
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyNetworkModeTransparent
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicy
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicy (411.53s)
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyComplete
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyNetworkModeTransparent (503.15s)
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicy
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicyComplete (560.39s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicy (667.14s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate (683.03s)
--- PASS: TestAccKubernetesCluster_addonProfileAzurePolicy (806.21s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyComplete (500.62s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureNPMPolicy (511.80s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate (1031.01s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate (1039.71s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containers	1039.872s
```

I had to add `node_resource_group = "${azurerm_resource_group.test.name}-nodes"` to each test.

Running tests in a location with a long name, e.g. `germanywestcentral`, results into an error:

```
        Resource Group Name: "acctestRG-aks-240531233634701724"
        Kubernetes Cluster Name: "acctestaks240531233634701724"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with response: {
          "code": "InvalidParameter",
          "details": null,
          "message": "The length of the node resource group name is too long. The maximum length is 80 and the length of the value provided is 83. Please see https://aka.ms/aks-naming-rules for more details.",
          "subcode": "",
          "target": "name"
         }
```

Had to use `germanywestcentral` because `westeurope` was full.

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_kubernetes_cluster` - Changing `network_profile.network_policy` does not force recreation.

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Comment by stephybun** [comment]
Thanks for this PR @jkroepke.

This feature in requested over in https://github.com/hashicorp/terraform-provider-azurerm/issues/25597 which contains an explanation for why we're [halting the addition of new preview features to the AKS resource](https://github.com/hashicorp/terraform-provider-azurerm/issues/25597#issuecomment-2114183649).

Given the explanation linked above we're unable to add support for this behaviour at this point in time so I'm going to close this PR for the time being. When a resolution is reached and/or the feature has come out of preview and is available in a stable API version we will be happy to re-open this. Thanks again for your effort and understanding on this!

**Comment by jkroepke** [comment]
@stephybun I disagree. If Issue you mention is about allow network_policy to "none",  and starts about an discussion about the API Version. However, this PR does not change the API version, nor changing the API version. It simply removing the ForceNew flag.

I can't understand, whats the connection between removing a ForeNew flag and #25597

**Comment by stephybun** [comment]
@jkroepke, allowing any kind of update on the `network_policy` field by removing the `ForceNew` on it, allows certain preview functionalities into the resource. 

Your tests check that `network_policy` can be removed, this I believe implies a `network_policy` of `none` which is the step required to uninstall Azure's Network Policy Manager or Calico - according to the documentation linked, that is a preview feature. 

https://learn.microsoft.com/en-us/azure/aks/use-network-policies#uninstall-azure-network-policy-manager-or-calico-preview

**Comment by jkroepke** [comment]
In summarize to unblock this PR, I can re-introduce ForceNew flag which triggers if is the new value is empty and the old not?

**Comment by jkroepke** [comment]
> this I believe implies a `network_policy` of `none`

I test that in an local test to validate that statement:

```hcl
func TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
	r := KubernetesClusterResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, ""),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, "azure"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").Exists(),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").HasValue("azure"),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, ""),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").Exists(),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").HasValue("azure"),
			),
		},
		data.ImportStep(),
	})
}

func TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
	r := KubernetesClusterResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, ""),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, "calico"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").Exists(),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").HasValue("calico"),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithOptionalPolicyConfig(data, ""),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").Exists(),
				check.That(data.ResourceName).Key("network_profile.0.network_policy").HasValue("calico"),
			),
		},
		data.ImportStep(),
	})
}
```

```
TF_ACC=1 go test -v ./internal/services/containers '-run=TestAccKubernetesCluster_advancedNetworking.+PolicyUpdate' -timeout 300m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
=== RUN   TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== PAUSE TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate
=== CONT  TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCiliumPolicyUpdate (614.03s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureAzurePolicyUpdate (1009.71s)
--- PASS: TestAccKubernetesCluster_advancedNetworkingAzureCalicoPolicyUpdate (1169.64s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containers	1169.753s
```

It turn out, omit the value would not uninstall the Network Policy. Uninstall the Network Policy requires an explicit `none`.

But I would agree that this may confuse end-users, too. I re-add the ForceNew condition which should deny that mention situation. But in any situation, a preview feature is not used in this PR.

**Comment by jkroepke** [comment]
> for needing to drill down into the details 

Sure, ask me what ever you need for your final decision. In mean time, can we re-open that PR please?

> It would help if you could explain the specific use case you're trying to support?

I have clusters which does not have a network_policy and I want to install one. Currently, only cillium is supported at the provider, but reading the docs, azure and calico are supported, too.

> If we're omitting the value sent, then what is the API sending back/what is the cluster falling back on for its network policy?

I did not dig too much into it, but if I omit the value and state still contains `azure`, I assume the old value is just returned from the API. I also observe, that the Azure Portal still displayed the network policy which I set before.

It seems like that omit the value means, it does not change the value.

File: internal/services/containers/kubernetes_cluster_addons_resource_test.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_addons_resource_test.go:341]: Please remove all of these, we run the acceptance tests on `westeurope` and I think france is the alternative location. Capacity issues are usually temporary anyway.

File: internal/services/containers/kubernetes_cluster_resource.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:116]: We can't remove this since this would inadvertently expose some preview functionality. We can extend this to allow updating from no network policy to azure or calico without recreating the cluster.

File: internal/services/containers/kubernetes_cluster_resource.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:2314]: This isn't necessary, since this is taken care of by the expand function for the `network_profile` block
```suggestion
```

File: internal/services/containers/kubernetes_cluster_network_resource_test.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_network_resource_test.go:261]: A test that validates omitting `network_policy` after it has been set is misleading in this instance because the property is `Computed`. 

The tests should validate that we can update an existing cluster with no network policy to having a network policy of `azure` and or `calico`.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for the details @jkroepke. Could you look over the comments left in-line? Once that's done we can take another look through.

File: internal/services/containers/kubernetes_cluster_addons_resource_test.go
Comment (jkroepke) [reply internal/services/containers/kubernetes_cluster_addons_resource_test.go:341]: Isn't it achievable that acceptance tests works in every region?

Comment (stephybun) [reply internal/services/containers/kubernetes_cluster_addons_resource_test.go:341]: Yes, but `westeurope` is where we have to run the acceptance tests and the tests should not include `Optional` properties that are not required for testing the related functionality. The limitation is coming from the AKS API so I would open an issue there to get that resolved.

Comment (jkroepke) [reply internal/services/containers/kubernetes_cluster_addons_resource_test.go:341]: If could cut 3 chars from `acctestRG-aks-240531233634701724`, it would also resolve the issue.

I will leave the change, until other issues are resolve, I will remove the lines, but I also do not want to re-add them all the times for local testing.

**Comment by jkroepke** [comment]
I apply the suggestions from you. Please take an other look.

File: internal/services/containers/kubernetes_cluster_network_resource_test.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_network_resource_test.go:283]:
```suggestion
func TestAccKubernetesCluster_advancedNetworkingAzurePolicyUpdate(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
	r := KubernetesClusterResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.advancedNetworkingConfig(data, "azure"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithPolicyConfig(data, "azure", "azure"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
	})
}
```

File: internal/services/containers/kubernetes_cluster_network_resource_test.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_network_resource_test.go:306]:
```suggestion
func TestAccKubernetesCluster_advancedNetworkingCalicoPolicyUpdate(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
	r := KubernetesClusterResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.advancedNetworkingConfig(data, "azure"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
		{
			Config: r.advancedNetworkingWithPolicyConfig(data, "azure", "calico"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
	})
}
```

File: internal/services/containers/kubernetes_cluster_network_resource_test.go
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_network_resource_test.go:359]: Removing the `network_policy` property from the config doesn't test any crucial behaviour since it's Computed so there is no need for this test at all.
```suggestion
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @jkroepke. Could you please address the comments left regarding the tests? Existing test configs can be re-used and several of the specific `Exists` and `HasValue` checks are taken care of by the `ImportStep` and can be removed.

Would you also mind updating the documentation?

**Review by stephybun — APPROVED** [review]
Thanks @jkroepke LGTM 👍

**Comment by jkroepke** [comment]
I really ask myself, why I did not had that behavior and what I did wrong.

**Comment by stephybun** [comment]
@jkroepke I ran the tests in TeamCity and they all passed, but tests that are supposed to check updates don't flag if the update causes the resource to be recreated. You can really only see that behaviour if you're looking at the API requests/responses that the provider is sending while the test is running, or by building the provider locally and running through the steps manually. 

This was in part a mistake on my end and one of the [review comments I left was also false](https://github.com/hashicorp/terraform-provider-azurerm/pull/26176#discussion_r1632940070). You did everything correct in that you followed instruction, I was not thorough enough in my review and too eager to have this merged, so the fault is with me. 

I'm looking at this in detail and will open a PR shortly to fix the behaviour.

**Comment by jkroepke** [comment]
What I have to configure that I see the output that you shows me? That would help me in the future to indicate if the behavior is correctly?

**Comment by stephybun** [comment]
I assume you're referring to the output in [this comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/26176#issuecomment-2157819126)?

This is just the output from running the test locally after I updated the `ForceNewIf` condition to not trigger a recreation when `network_policy` was updated.

If you're asking how to view the API requests/responses from the tests, this can't be configured as output in the tests. You either need to check the debug logs, or be running the tests with a proxy that can capture the traffic. I hope that answers your questions.

**Pull request merged**
